File: src/main/java/com/github/zyypj/tadeuBooter/api/file/YAML.java
package com.github.zyypj.tadeuBooter.api.file;

import org.bukkit.ChatColor;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.Plugin;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class YAML extends YamlConfiguration {

    private final Plugin plugin;
    private final File configFile;

    public YAML(String name, Plugin plugin) throws IOException, InvalidConfigurationException {
        this(name, plugin, plugin.getDataFolder());
    }

    public YAML(String name, Plugin plugin, File folder) throws IOException, InvalidConfigurationException {
        this.plugin = plugin;
        if (folder == null) folder = plugin.getDataFolder();
        if (!folder.exists()) {
            boolean created = folder.mkdirs();
            if (!created) {
                throw new IOException("Could not create folder: " + folder.getAbsolutePath());
            }
        }

        this.configFile = new File(folder, name.endsWith(".yml") ? name : name + ".yml");
        this.loadConfig();
    }

    private void loadConfig() throws IOException, InvalidConfigurationException {
        if (!this.configFile.exists()) {
            if (plugin.getResource(this.configFile.getName()) != null) {
                copyResource(this.configFile.getName(), this.configFile);
            } else {
                boolean created = this.configFile.createNewFile();
                if (!created) {
                    throw new IOException("Could not create config file: " + configFile.getAbsolutePath());
                }
            }
        }
        this.load(this.configFile);
    }

    private void copyResource(String resourcePath, File destination) throws IOException {
        try (InputStream in = plugin.getResource(resourcePath)) {
            if (in == null) {
                throw new IOException("Resource " + resourcePath + " not found in plugin jar");
            }
            Files.copy(in, destination.toPath(), StandardCopyOption.REPLACE_EXISTING);
        }
    }

    public void saveDefaultConfig() {
        if (!this.configFile.exists()) {
            if (plugin.getResource(this.configFile.getName()) != null) {
                try {
                    copyResource(this.configFile.getName(), this.configFile);
                } catch (IOException e) {
                    logError("Erro ao salvar recurso padrão para o arquivo " + this.configFile.getName(), e);
                }
            } else {
                try {
                    if (!this.configFile.getParentFile().exists()) {
                        boolean created = this.configFile.getParentFile().mkdirs();
                        if (!created) {
                            logError("Não foi possível criar diretório para o arquivo " + this.configFile.getName(), new IOException("mkdirs retornou false"));
                        }
                    }
                    boolean created = this.configFile.createNewFile();
                    if (!created) {
                        logError("Não foi possível criar o arquivo " + this.configFile.getName(), new IOException("createNewFile retornou false"));
                    }
                } catch (IOException e) {
                    logError("Erro ao criar o arquivo " + this.configFile.getName(), e);
                }
            }
        }
    }

    public void createDefaults() {
        saveDefaultConfig();
    }

    public boolean exists() {
        return this.configFile.exists();
    }

    public boolean delete() {
        return this.configFile.delete();
    }

    public void backup(String suffix) {
        File backupFile = new File(configFile.getParent(), configFile.getName() + "." + suffix + ".backup");
        try {
            Files.copy(configFile.toPath(), backupFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
        } catch (IOException e) {
            plugin.getLogger().warning("Não foi possível criar o backup de: " + configFile.getName());
        }
    }

    public void save() {
        try {
            this.save(this.configFile);
        } catch (IOException e) {
            logError("Erro ao salvar o arquivo " + this.configFile.getName(), e);
        }
    }

    public void reload() {
        try {
            this.loadConfig();
        } catch (IOException | InvalidConfigurationException e) {
            logError("Erro ao recarregar o arquivo " + this.configFile.getName(), e);
        }
    }

    public void set(String path, Object value, boolean save) {
        this.set(path, value);
        if (save) this.save();
    }

    public void setDefault(String path, Object value) {
        if (!this.contains(path)) {
            this.set(path, value);
            this.save();
        }
    }

    public String getString(String path, boolean translateColors) {
        return getString(path, null, translateColors);
    }

    public String getString(String path, String defaultValue, boolean translateColors) {
        String value = this.getString(path, defaultValue);
        return value != null && translateColors ? ChatColor.translateAlternateColorCodes('&', value) : value;
    }

    public List<String> getStringList(String path, boolean translateColors) {
        List<String> list = this.getStringList(path);
        if (translateColors && list != null) {
            return list.stream()
                    .map(line -> ChatColor.translateAlternateColorCodes('&', line))
                    .collect(Collectors.toList());
        }
        return list;
    }

    public <T> T getOrSet(String path, T defaultValue) {
        if (!this.contains(path)) {
            this.set(path, defaultValue);
            this.save();
        }
        return (T) this.get(path);
    }

    public String getFormatted(String path, Object... args) {
        String raw = getString(path, true);
        return raw != null ? String.format(raw, args) : null;
    }

    public List<String> getKeysRecursive(String path) {
        org.bukkit.configuration.ConfigurationSection section = this.getConfigurationSection(path);
        if (section == null) return Collections.emptyList();
        return new ArrayList<>(section.getKeys(true));
    }

    private void logError(String msg, Throwable t) {
        plugin.getLogger().severe("§c[YAML] " + msg);
        t.printStackTrace();
    }
}

File: spigot/src/main/java/me/zyypj/booter/minecraft/spigot/file/YAML.kt
package me.zyypj.booter.minecraft.spigot.file

import org.bukkit.ChatColor
import org.bukkit.configuration.InvalidConfigurationException
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.plugin.Plugin
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption

/**
 * Extensão de YamlConfiguration que facilita criação, carregamento e manipulação de arquivos YML.
 * Construtores com @JvmOverloads disponíveis para uso em Java.
 */
class YAML @JvmOverloads constructor(
    name: String,
    private val plugin: Plugin,
    folder: File? = null
) : YamlConfiguration() {

    private val configFile: File

    init {
        val dir = folder ?: plugin.dataFolder
        if (!dir.exists() && !dir.mkdirs()) {
            throw IOException("Could not create folder: ${dir.absolutePath}")
        }
        val fileName = if (name.endsWith(".yml")) name else "$name.yml"
        configFile = File(dir, fileName)
        loadConfig()
    }

    @Throws(IOException::class, InvalidConfigurationException::class)
    private fun loadConfig() {
        if (!configFile.exists()) {
            val resource = plugin.getResource(configFile.name)
            if (resource != null) {
                copyResource(configFile.name, configFile)
            } else if (!configFile.createNewFile()) {
                throw IOException("Could not create config file: ${configFile.absolutePath}")
            }
        }
        load(configFile)
    }

    @Throws(IOException::class)
    private fun copyResource(resourcePath: String, destination: File) {
        plugin.getResource(resourcePath)?.use { input ->
            Files.copy(input, destination.toPath(), StandardCopyOption.REPLACE_EXISTING)
        } ?: throw IOException("Resource $resourcePath not found in plugin jar")
    }

    /** Salva recurso ‘default’ se não existir. */
    fun saveDefaultConfig() {
        if (!configFile.exists()) {
            plugin.getResource(configFile.name)?.let {
                try {
                    copyResource(configFile.name, configFile)
                } catch (e: IOException) {
                    logError("Erro ao salvar recurso padrão para o arquivo ${configFile.name}", e)
                }
            } ?: run {
                try {
                    configFile.parentFile?.let { parent ->
                        if (!parent.exists() && !parent.mkdirs())
                            logError(
                                "Não foi possível criar diretório para o arquivo ${configFile.name}",
                                IOException("mkdirs retornou false")
                            )
                    }
                    if (!configFile.createNewFile())
                        logError(
                            "Não foi possível criar o arquivo ${configFile.name}",
                            IOException("createNewFile retornou false")
                        )
                } catch (e: IOException) {
                    logError("Erro ao criar o arquivo ${configFile.name}", e)
                }
            }
        }
    }

    /** Alias de saveDefaultConfig. */
    fun createDefaults() = saveDefaultConfig()

    /** Verifica existência do arquivo. */
    fun exists(): Boolean = configFile.exists()

    /** Deleta o arquivo. */
    fun delete(): Boolean = configFile.delete()

    /** Cria backup com sufixo. */
    fun backup(suffix: String) {
        val backup = File(configFile.parent, "${configFile.name}.$suffix.backup")
        try {
            Files.copy(configFile.toPath(), backup.toPath(), StandardCopyOption.REPLACE_EXISTING)
        } catch (e: IOException) {
            plugin.logger.warning("Não foi possível criar o backup de: ${configFile.name}")
        }
    }

    /** Salva alterações no arquivo. */
    fun save() {
        try {
            save(configFile)
        } catch (e: IOException) {
            logError("Erro ao salvar o arquivo ${configFile.name}", e)
        }
    }

    /** Recarrega do disco. */
    fun reload() {
        try {
            loadConfig()
        } catch (e: Exception) {
            logError("Erro ao recarregar o arquivo ${configFile.name}", e)
        }
    }

    /** Ajusta valor em path e salva se indicado. */
    fun set(path: String, value: Any?, save: Boolean) {
        set(path, value)
        if (save) save()
    }

    /** Define default, se não existir, e salva. */
    fun setDefault(path: String, value: Any) {
        if (!contains(path)) {
            set(path, value)
            save()
        }
    }

    /** Obtém string, opcionalmente traduzindo cores `&`. */
    fun getString(path: String, translateColors: Boolean): String? {
        val valRaw = getString(path)
        return if (translateColors && valRaw != null)
            ChatColor.translateAlternateColorCodes('&', valRaw)
        else valRaw
    }

    /** Lista de strings, opcionalmente traduzindo cores. */
    fun getStringList(path: String, translateColors: Boolean): List<String> {
        val list = getStringList(path)
        return if (translateColors)
            list.map { ChatColor.translateAlternateColorCodes('&', it) }
        else list
    }

    /** Retorna T e seta default se não existe. */
    @Suppress("UNCHECKED_CAST")
    fun <T> getOrSet(path: String, default: T): T {
        if (!contains(path)) {
            set(path, default)
            save()
        }
        return get(path) as T
    }

    /** Formata string usando String.format. */
    fun getFormatted(path: String, vararg args: Any): String? {
        val raw = getString(path, true) ?: return null
        return String.format(raw, *args)
    }

    /** Chaves recursivas dentro de uma seção. */
    fun getKeysRecursive(path: String): List<String> {
        val section = getConfigurationSection(path) ?: return emptyList()
        return section.getKeys(true).toList()
    }

    private fun logError(msg: String, t: Throwable) {
        plugin.logger.severe("§c[YAML] $msg")
        t.printStackTrace()
    }
}